code/controllers/ShareDraftController.php

```diff
@@ -42,25 +42,50 @@ public function preview(SS_HTTPRequest $request) {
 			return $this->errorPage();
 		}
 
-		$page = $shareToken->Page();
+		$page = Versioned::get_one_by_stage(
+			'SiteTree',
+			'Stage',
+			sprintf('"SiteTree"."ID" = \'%d\'', $shareToken->PageID)
+		);
 
-		$latest = $page->Versions(null, 'Version DESC')->first();
+		$latest = Versioned::get_latest_version('SiteTree', $shareToken->PageID);
 
-		$controller = $this->getControllerFor($latest);
+		$controller = $this->getControllerFor($page);
 
 		if(!$shareToken->isExpired() && $page->generateKey($shareToken->Token) === $key) {
 			Requirements::css(SHAREDRAFTCONTENT_DIR . '/css/top-bar.css');
 
-			$rendered = $controller->render();
-
-			$data = new ArrayData(array(
-				'Page' => $page,
-				'Latest' => $latest,
-			));
-
-			$include = (string) $data->renderWith('Includes/TopBar');
-
-			return str_replace('</body>', $include . '</body>', (string) $rendered);
+			// Temporarily un-secure the draft site and switch to draft
+			$oldSecured = Session::get('unsecuredDraftSite');
+			$oldMode = Versioned::get_reading_mode();
+			$restore = function() use ($oldSecured, $oldMode) {
+				Session::set('unsecuredDraftSite', $oldSecured);
+				Versioned::set_reading_mode($oldMode);
+			};
+
+			// Process page inside an unsecured draft container
+			try {
+				Session::set('unsecuredDraftSite', true);
+				Versioned::reading_stage('Stage');
+
+				// Create mock request; Simplify request to single top level reqest
+				$pageRequest = new SS_HTTPRequest('GET', $page->URLSegment);
+				$pageRequest->match('$URLSegment//$Action/$ID/$OtherID', true);
+				$rendered = $controller->handleRequest($pageRequest, $this->model);
+
+				// Render draft heading
+				$data = new ArrayData(array(
+					'Page' => $page,
+					'Latest' => $latest,
```

Comment:
I'm not really sure separate $latest and $page is needed, but I've left in $latest anyway.